### PR TITLE
[branching io manager][fix] Fix loading partitioned parent assets in BranchingIOManager

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Tweak of fix in #17303 which uses the upstream asset partition rather than the current asset partition when using the branching IO manager. Uses `context.asset_partition_key` rather than the upstream output's partition key, which can sometimes be `None` when the upstream output is not available.

## Test Plan

Adds a new unit test with a non-standard partition mapping, which previously failed.

### CODE REVIEW

**Comment by benpankow** [comment]
Current dependencies on/for this PR:
* master
  * **PR #18491** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/18491?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/18491?utm_source=stack-comment).

**Comment by smackesey** [comment]
I'm confused as to why `InputContext.asset_partition_key` and `InputContext.partition_key` differ in the first place. Shouldn't the `InputContext.partition_key` be scoped to the input being loaded instead of the run? Do we ever use the run value?

**Comment by danielgafni** [comment]
We might not be loading an asset at all. The IOManager should also work for normal jobs.

**Comment by smackesey** [comment]
>We might not be loading an asset at all. The IOManager should also work for normal jobs. This is because it’s often used as a drop-in replacement for an existing IOManager.

That's true, but the question I'm raising is about `InputContext.partition_key`. Seems to me this should mean "partition key for the input that this `InputContext` is defined for". If that were true it would not interfere with the ability of any IO manager to work with non-assets.

**Comment by benpankow** [comment]
>  I'm confused as to why InputContext.asset_partition_key and InputContext.partition_key differ in the first place. Shouldn't the InputContext.partition_key be scoped to the input being loaded instead of the run? Do we ever use the run value?

I'm not sure why it's set to be the partition for the asset that's currently being materialized.

It is pretty confusing especially since the two properties would seemingly be the same thing - the input manager should really be agnostic to what's being produced in the actual compute step.

>  We might not be loading an asset at all. The IOManager should also work for normal jobs. This is because it’s often used as a drop-in replacement for an existing IOManager.

This is a good call-out - right now if there's no upstream asset involved the IO manager will fallback to `None` partition, so it should continue to work in the op case. Will add a test just to sanity check.

**Review by alangenfeld — COMMENTED** [review]
This touches on the complexity around how partitioning works and is modeled that I don't have a great grasp on. 

I don't think its OK to make a breaking change the `@public` `partition_key` method, until 2.0 given my understanding of our policy. My first reaction was that maybe we could deprecate it and replace it with `run_partition_key` to more clearly disambiguate, but I don't have confidence that thats the right move. 

When does the upstream output partition key differ from the run partition key in the non asset case? From what I understand you can only attach `PartitionMapping` to assets.

File: python_modules/dagster/dagster/_core/storage/branching/branching_io_manager.py [hunk 1]
```diff
@@ -69,10 +70,20 @@ def load_input(self, context: InputContext) -> Any:
             return self.branch_io_manager.load_input(context)
         else:
             # we are dealing with an asset input
+            partition_key = (
+                context.upstream_output.partition_key
+                if context.upstream_output and context.upstream_output.has_partition_key
+                else None
+            )
+            try:
+                partition_key = context.asset_partition_key
```
[the remark this answers is not in the document]
Comment (benpankow) [reply python_modules/dagster/dagster/_core/storage/branching/branching_io_manager.py:79]: Updated logic & added test case, should be able to handle strange partition mappings now.

**Comment by smackesey** [comment]
>When does the upstream output partition key differ from the run partition key in the non asset case? From what I understand you can only attach PartitionMapping to assets.

Pretty sure it doesn't. So it would only change in the asset case. I am supportive of introducing `run_partition_key` and changing `partition_key` to reflect the input in 2.0 (with a warning introduced in the meantime, that we can emit only if we are loading an asset).

File: python_modules/dagster/dagster/_core/storage/branching/branching_io_manager.py [hunk 2]
```diff
@@ -69,18 +69,30 @@ def load_input(self, context: InputContext) -> Any:
...
+
+            event_log_entries = [
+                latest_materialization_log_entry(
+                    instance=context.instance,
+                    asset_key=context.asset_key,
+                    partition_key=partition_key,
+                )
+                for partition_key in partition_keys
+            ]
+            if all(
```
[the remark this answers is not in the document]
Comment (benpankow) [reply python_modules/dagster/dagster/_core/storage/branching/branching_io_manager.py:88]: https://github.com/dagster-io/dagster/pull/18491/files/7bee3fd9cdf04c032e6955d5eadd05d5c17b7729#r1422995242 should ensure that the length is always >=1, will add a note though

**Pull request merged**